File: velinor/game_mechanics/characters.ts
// TypeScript version of simple game mechanics for Velinor.
// Minimal, framework-agnostic classes for use in frontend code.

export type ToneKeys = 'trust' | 'observation' | 'narrative' | 'empathy';

export interface ToneState {
  trust: number;
  observation: number;
  narrative: number;
  empathy: number;
}

export class Tone implements ToneState {
  trust = 0.5;
  observation = 0.5;
  narrative = 0.5;
  empathy = 0.5;

  clamp() {
    (['trust', 'observation', 'narrative', 'empathy'] as ToneKeys[]).forEach(k => {
      // @ts-ignore
      this[k] = Math.max(0, Math.min(1, this[k]));
    });
  }

  asObject(): ToneState {
    return { trust: this.trust, observation: this.observation, narrative: this.narrative, empathy: this.empathy };
  }
}

export class GameMechanics {
  data: any;
  tone: Tone;
  attunement: number;

  constructor(data: any) {
    this.data = data;
    this.tone = new Tone();
    this.attunement = 0;
  }

  applyChoice(orientation: string, choiceIndex: number) {
    const scene = this.data.miniGames.observationalScene;
    const order: ToneKeys[] = scene.toneOrder;
    const shifts: number[] = scene.toneShifts[orientation][choiceIndex];
    order.forEach((k, i) => {
      // @ts-ignore
      this.tone[k] += shifts[i];
    });
    this.tone.clamp();
    this.attunement += shifts[3] * 0.5; // empathy-weighted increment
  }

  malrikElenyaCorrelation(): number {
    const thresholds = this.data.systems.attunement.thresholds;
    const cfg = this.data.systems.npcCorrelation.malrik_elenya;
    if (this.attunement <= thresholds[0]) return cfg.default;
    if (this.attunement <= thresholds[1]) return cfg.neutral;
    return cfg.positive;
  }
}

// Example usage (import JSON as data and create GameMechanics)
// import data from '../markdowngameinstructions/malrik_elenya_coren.json'
// const gm = new GameMechanics(data);
// gm.applyChoice('empathy', 0);
// console.log(gm.tone.asObject(), gm.attunement, gm.malrikElenyaCorrelation());

File: velinor/game_mechanics/characters.py
"""
Simple mechanics implementation for Velinor characters and tone system.
This module loads the JSON export and provides small classes to apply mini-game choices
and update player tone stats and NPC correlations.
"""
from dataclasses import dataclass, field
from typing import Dict, List
import json
import os


ROOT = os.path.dirname(__file__)  # velinor/game_mechanics
DATA_PATH = os.path.join(os.path.dirname(ROOT), 'markdowngameinstructions', 'malrik_elenya_coren.json')


def load_data(path: str = DATA_PATH) -> Dict:
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


@dataclass
class ToneState:
    trust: float = 0.5
    observation: float = 0.5
    narrative: float = 0.5
    empathy: float = 0.5

    def clamp(self):
        for k in ('trust', 'observation', 'narrative', 'empathy'):
            v = getattr(self, k)
            setattr(self, k, max(0.0, min(1.0, v)))

    def as_dict(self) -> Dict[str, float]:
        return {k: getattr(self, k) for k in ('trust', 'observation', 'narrative', 'empathy')}


@dataclass
class GameMechanics:
    data: Dict = field(default_factory=load_data)
    tone: ToneState = field(default_factory=ToneState)
    attunement: float = 0.0

    def apply_choice(self, orientation: str, choice_index: int):
        scene = self.data['miniGames']['observationalScene']
        order = scene['toneOrder']
        shifts = scene['toneShifts'][orientation][choice_index]
        # shifts array corresponds to order [trust, observation, narrative, empathy]
        for key, delta in zip(order, shifts):
            setattr(self.tone, key, getattr(self.tone, key) + delta)
        self.tone.clamp()
        # update attunement: empathy-weighted increment (simplified)
        self.attunement += shifts[3] * 0.5
        self.attunement = max(0.0, self.attunement)

    def malrik_elenya_correlation(self) -> float:
        # Determine correlation state based on attunement thresholds
        thresholds = self.data['systems']['attunement']['thresholds']
        if self.attunement <= thresholds[0]:
            return self.data['systems']['npcCorrelation']['malrik_elenya']['default']
        if self.attunement <= thresholds[1]:
            return self.data['systems']['npcCorrelation']['malrik_elenya']['neutral']
        return self.data['systems']['npcCorrelation']['malrik_elenya']['positive']


if __name__ == '__main__':
    gm = GameMechanics()
    print('Initial tone:', gm.tone.as_dict())
    print('Initial attunement:', gm.attunement)
    # Example: player with empathy orientation picks choice 0 (strong empathy)
    gm.apply_choice('empathy', 0)
    print('After empathy choice:', gm.tone.as_dict())
    print('Attunement:', gm.attunement)
    print('Malrik/Elenya correlation:', gm.malrik_elenya_correlation())
